### ISSUE TITLE
Iteratively Improve Model Based on Feedback

### ISSUE THREAD

**Issue by judejudo** [openchlai/ai#52]
# Git Task: Iterative Model Improvement

## Description:
Refine the initial prediction model based on insights from the error analysis. This may involve adding new features (e.g., TF-IDF, embeddings), addressing class imbalance, or experimenting with model architecture and hyperparameters. Track all changes to ensure reproducibility and improvement traceability.

---

## Acceptance Criteria:
- **Performance Improvement:**
  - [x] Model accuracy on the validation set improves by **≥5%** compared to model_v1.
  
- **Change Documentation:**
  - [ ] All modifications (e.g., feature additions, data adjustments, architecture changes) are clearly documented in a **model_updates.md** changelog.
  
- **Reproducibility:**
  - [ ] All code changes and model updates are tracked with version control (e.g., Git, DVC, MLflow).

- **Deliverables:**
  - [ ] **model_v2.pkl**: Updated model with improved performance.
  - [ ] **model_updates.md**: Document listing all modifications, rationale, and impact on model performance.

---

## Tasks:
- [ ] **Review Error Analysis Insights:**
  - Identify common failure patterns from `error_analysis.ipynb`.
  - Prioritize improvement strategies based on error trends (e.g., false negatives on "violence" cases).

- [ ] **Add/Engineer New Features:**
  - Implement new features such as:
    - [ ] TF-IDF vectors.
    - [ ] Additional linguistic features (e.g., part-of-speech tags, named entity counts).
    - [ ] Contextual embeddings (e.g., BERT tokens).
  - [ ] Evaluate impact of new features using cross-validation.

- [ ] **Balance Training Data:**
  - [ ] Apply oversampling/undersampling techniques if class imbalance is observed.
  - [ ] Evaluate the effect of balanced data on model metrics.

- [ ] **Update Model Architecture or Algorithm (if needed):**
  - [ ] Fine-tune hyperparameters.
  - [ ] Try alternative models (e.g., switch from XGBoost to BERT or vice versa).
  - [ ] Optimize for generalization (e.g., dropout, regularization).

- [ ] **Retrain Improved Model:**
  - [ ] Train model with updated features/data/architecture.
  - [ ] Evaluate updated model performance.
  - [ ] Ensure accuracy improves by at least 5%.

- [ ] **Save Updated Model:**
  - [ ] Save as `model_v2.pkl` and track with DVC or MLflow.

- [ ] **Document Improvements:**
  - [ ] Create `model_updates.md` with:
    - Changes made.
    - Rationale for each change.
    - Performance comparison with previous version.
    - Metrics table showing improvements.

---

## Dependencies:
- Completion of **Issue 3: Evaluate Model Performance** (including error analysis).

## Effort Estimate:
- **2 weeks** (ML Engineer + Data Scientist).

**Comment by nelsonadagi** [comment]
@K-NURF @shemmiriam @ogutubrenda @marlonbino @Rogendo 

🔹 Automated Workflow for Iteration Logbook → Stakeholder Report
1. Maintain Raw Evidence in Git

Keep everything in your repo under folders:

/logs/ITERATION_LOG.md
/metrics/iteration_X_results.json
/feedback/iteration_X_feedback.md
/reports/templates/

2. Create a Report Template (Markdown → PDF/Docx)

Inside reports/templates/iteration_report_template.md:

# Iteration Summary Report – {{ date }}

## 1. Overview
- Iteration cycle: {{ iteration }}
- Purpose: {{ purpose }}

## 2. Before/After Metrics
| Metric      | Before | After |
|-------------|--------|-------|
{{ metrics_table }}

## 3. Feedback Highlights
{{ feedback_summary }}

## 4. Confusion Matrix
![Confusion Matrix](../evidence/{{ confusion_matrix }})

## 5. Annex
- Full Iteration Log: see `logs/ITERATION_LOG.md`
- Raw evaluation report: `metrics/{{ evaluation_file }}`

3. Automate Report Generation Script (Python)

Create generate_report.py in /scripts/:

import json
from datetime import date
import pypandoc

# Load metrics from JSON
with open("metrics/iteration_3_results.json") as f:
    results = json.load(f)

# Create metrics table
metrics_table = ""
for metric, values in results["before_after"].items():
    metrics_table += f"| {metric} | {values['before']} | {values['after']} |\n"

# Load feedback summary
with open("feedback/iteration_3_feedback.md") as f:
    feedback_summary = f.read()

# Fill in template
with open("reports/templates/iteration_report_template.md") as f:
    template = f.read()

report_md = template.replace("{{ date }}", str(date.today())) \
    .replace("{{ iteration }}", "Iteration 3") \
    .replace("{{ purpose }}", "Improved recall via class balancing") \
    .replace("{{ metrics_table }}", metrics_table) \
    .replace("{{ feedback_summary }}", feedback_summary) \
    .replace("{{ confusion_matrix }}", "confusion_matrix_iter3.png") \
    .replace("{{ evaluation_file }}", "iteration_3_results.json")

# Save Markdown
with open("reports/iteration_3_summary.md", "w") as f:
    f.write(report_md)

# Convert to PDF
pypandoc.convert_text(report_md, "pdf", format="md",
    outputfile="reports/iteration_3_summary.pdf",
    extra_args=["--standalone"])

4. Automate with GitHub Actions (CI/CD)

In .github/workflows/report.yml:

name: Generate Iteration Reports

on:
  push:
    paths:
      - "logs/**"
      - "metrics/**"
      - "feedback/**"

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3

      - name: Set up Python
        uses: actions/setup-python@v4
        with:
          python-version: "3.10"

      - name: Install deps
        run: pip install pypandoc

      - name: Generate Report
        run: python scripts/generate_report.py

      - name: Upload Artifact
        uses: actions/upload-artifact@v3
        with:
          name: iteration-report
          path: reports/iteration_*.pdf

5. Distribution

Every time you push new logs/metrics/feedback → GitHub Actions auto-generates a PDF report.

You can:

Email the artifact automatically (via GitHub Action + SMTP).

Upload to Google Drive / SharePoint.

Keep them in a /stakeholder-reports/ folder that syncs automatically.
